File: apps/ios/MagicMobile/GameStats.swift
import Foundation

/// What happened in one game, gathered from consecutive snapshots for the result screen.
/// Only public information is used: life totals, combat groups and zone contents.
struct GameStats: Equatable {
    private(set) var gameID = ""
    private(set) var turns = 0
    private(set) var startingLife: Int?
    private(set) var finalLife: Int?
    /// Combat damage your unblocked attackers dealt to players.
    private(set) var combatDamage = 0
    /// The largest combat hit you landed on one player at once.
    private(set) var biggestHit = 0
    private(set) var damageByCard: [String: Int] = [:]
    private(set) var creaturesDestroyed = 0
    private(set) var creaturesLost = 0
    private var previous: Capture?

    /// The lightweight part of a snapshot the next one is compared with.
    private struct Capture: Equatable {
        let life: [String: Int]
        let battlefieldCreatures: [String: Set<String>]
        let attacks: [Attack]
    }

    private struct Attack: Equatable {
        let defender: String
        let attackers: [(name: String, power: Int)]

        static func == (lhs: Attack, rhs: Attack) -> Bool {
            lhs.defender == rhs.defender && lhs.attackers.map(\.name) == rhs.attackers.map(\.name)
                && lhs.attackers.map(\.power) == rhs.attackers.map(\.power)
        }
    }

    mutating func record(_ snapshot: GameSnapshot) {
        if snapshot.id != gameID { self = GameStats(); gameID = snapshot.id }
        turns = max(turns, snapshot.turn)
        let viewer = snapshot.viewerID
        if let life = snapshot.human?.life {
            if startingLife == nil { startingLife = life }
            finalLife = life
        }
        let yourBattlefield = Set(snapshot.human?.zones.battlefield.map(\.instanceId) ?? [])
        let capture = Capture(
            life: Dictionary(snapshot.players.map { ($0.playerId, $0.life) }, uniquingKeysWith: { first, _ in first }),
            battlefieldCreatures: Dictionary(snapshot.players.map { player in
                (player.playerId, Set(player.zones.battlefield.filter(\.isCreature).map(\.instanceId)))
            }, uniquingKeysWith: { first, _ in first }),
            attacks: (snapshot.xmage?.combat ?? []).compactMap { group in
                guard !group.blocked, group.defenderId != viewer else { return nil }
                let mine = group.attackers.filter { yourBattlefield.contains($0.instanceId) }
                    .map { (name: $0.card.name, power: max(0, Int($0.displayPower ?? "") ?? $0.power ?? 0)) }
                    .filter { $0.power > 0 }
                return mine.isEmpty ? nil : Attack(defender: group.defenderId, attackers: mine)
            }
        )
        if let previous { compare(previous, capture, snapshot: snapshot) }
        previous = capture
    }

    private mutating func compare(_ old: Capture, _ new: Capture, snapshot: GameSnapshot) {
        // Credit a defender's life loss to the unblocked attackers that were swinging at them.
        for attack in old.attacks {
            guard let before = old.life[attack.defender], let after = new.life[attack.defender], after < before else { continue }
            let total = attack.attackers.reduce(0) { $0 + $1.power }
            let credited = min(total, before - after)
            guard total > 0, credited > 0 else { continue }
            combatDamage += credited
            biggestHit = max(biggestHit, credited)
            for attacker in attack.attackers {
                damageByCard[attacker.name, default: 0] += Int((Double(attacker.power) * Double(credited) / Double(total)).rounded())
            }
        }
        // A creature that left the battlefield for its owner's graveyard died (tokens just vanish).
        for player in snapshot.players {
            let gone = (old.battlefieldCreatures[player.playerId] ?? []).subtracting(new.battlefieldCreatures[player.playerId] ?? [])
            guard !gone.isEmpty else { continue }
            let graveyard = Set(player.zones.graveyard.map(\.instanceId))
            let died = gone.intersection(graveyard).count
            if snapshot.isViewer(player.playerId) { creaturesLost += died } else { creaturesDestroyed += died }
        }
    }

    /// The card of yours that dealt the most combat damage to players.
    var topCard: (name: String, damage: Int)? {
        damageByCard.max { $0.value < $1.value || ($0.value == $1.value && $0.key > $1.key) }
            .flatMap { $0.value > 0 ? ($0.key, $0.value) : nil }
    }

    static func == (lhs: GameStats, rhs: GameStats) -> Bool {
        lhs.gameID == rhs.gameID && lhs.turns == rhs.turns && lhs.combatDamage == rhs.combatDamage
            && lhs.finalLife == rhs.finalLife && lhs.creaturesLost == rhs.creaturesLost
            && lhs.creaturesDestroyed == rhs.creaturesDestroyed && lhs.damageByCard == rhs.damageByCard
    }
}

File: apps/android/core/src/main/kotlin/io/magicmobile/android/game/GameStats.kt
package io.magicmobile.android.game

import kotlin.math.roundToInt

/**
 * Port of GameStats.swift: what happened in one game, gathered from consecutive snapshots for
 * the result screen. Only public information is used: life totals, combat groups and zones.
 */
class GameStats {
    var gameID = ""; private set
    var turns = 0; private set
    var startingLife: Int? = null; private set
    var finalLife: Int? = null; private set
    /** Combat damage your unblocked attackers dealt to players. */
    var combatDamage = 0; private set
    /** The largest combat hit you landed on one player at once. */
    var biggestHit = 0; private set
    val damageByCard = LinkedHashMap<String, Int>()
    var creaturesDestroyed = 0; private set
    var creaturesLost = 0; private set
    private var previous: Capture? = null

    private data class Attack(val defender: String, val attackers: List<Pair<String, Int>>)
    private data class Capture(val life: Map<String, Int>, val battlefieldCreatures: Map<String, Set<String>>, val attacks: List<Attack>)

    fun record(snapshot: GameSnapshot) {
        if (snapshot.id != gameID) reset(snapshot.id)
        turns = maxOf(turns, snapshot.turn)
        val viewer = snapshot.viewerID
        snapshot.human?.life?.let { life ->
            if (startingLife == null) startingLife = life
            finalLife = life
        }
        val yours = snapshot.human?.zones?.battlefield?.map { it.instanceId }?.toSet() ?: emptySet()
        val life = LinkedHashMap<String, Int>(); snapshot.players.forEach { life.putIfAbsent(it.playerId, it.life) }
        val creatures = LinkedHashMap<String, Set<String>>()
        snapshot.players.forEach { player -> creatures.putIfAbsent(player.playerId, player.zones.battlefield.filter { it.isCreature }.map { it.instanceId }.toSet()) }
        val attacks = (snapshot.xmage?.combat ?: emptyList()).mapNotNull { group ->
            if (group.blocked || group.defenderId == viewer) return@mapNotNull null
            val mine = group.attackers.filter { it.instanceId in yours }
                .map { it.card.name to maxOf(0, it.displayPower?.toIntOrNull() ?: it.power ?: 0) }
                .filter { it.second > 0 }
            if (mine.isEmpty()) null else Attack(group.defenderId, mine)
        }
        val capture = Capture(life, creatures, attacks)
        previous?.let { compare(it, capture, snapshot) }
        previous = capture
    }

    private fun reset(id: String) {
        gameID = id; turns = 0; startingLife = null; finalLife = null; combatDamage = 0; biggestHit = 0
        damageByCard.clear(); creaturesDestroyed = 0; creaturesLost = 0; previous = null
    }

    private fun compare(old: Capture, new: Capture, snapshot: GameSnapshot) {
        // Credit a defender's life loss to the unblocked attackers that were swinging at them.
        for (attack in old.attacks) {
            val before = old.life[attack.defender] ?: continue
            val after = new.life[attack.defender] ?: continue
            if (after >= before) continue
            val total = attack.attackers.sumOf { it.second }
            val credited = minOf(total, before - after)
            if (total <= 0 || credited <= 0) continue
            combatDamage += credited
            biggestHit = maxOf(biggestHit, credited)
            for ((name, power) in attack.attackers) {
                damageByCard[name] = (damageByCard[name] ?: 0) + (power.toDouble() * credited / total).roundToInt()
            }
        }
        // A creature that left the battlefield for its owner's graveyard died (tokens just vanish).
        for (player in snapshot.players) {
            val gone = (old.battlefieldCreatures[player.playerId] ?: emptySet()) - (new.battlefieldCreatures[player.playerId] ?: emptySet())
            if (gone.isEmpty()) continue
            val graveyard = player.zones.graveyard.map { it.instanceId }.toSet()
            val died = gone.count { it in graveyard }
            if (snapshot.isViewer(player.playerId)) creaturesLost += died else creaturesDestroyed += died
        }
    }

    /** The card of yours that dealt the most combat damage to players. */
    val topCard: Pair<String, Int>? get() = damageByCard.entries
        .maxWithOrNull { a, b -> if (a.value != b.value) a.value.compareTo(b.value) else b.key.compareTo(a.key) }
        ?.takeIf { it.value > 0 }?.let { it.key to it.value }
}

/** Port of OpeningHandChoice: XMage's mulligan question, answered from the opening-hand screen. */
class OpeningHandChoice private constructor(val message: String, val keepLabel: String, val mulliganLabel: String, val keep: GameCommand,
                                            val mulligan: GameCommand, val promptID: String) {
    companion object {
        private val tags = Regex("<[^>]+>")

        fun of(snapshot: GameSnapshot): OpeningHandChoice? {
            val prompt = snapshot.promptEnvelopeV2 ?: return null
            val confirmation = prompt.confirmation ?: return null
            if (!prompt.message.contains("mulligan", ignoreCase = true)) return null
            val yes = confirmation.yesCommand ?: return null
            val no = confirmation.noCommand ?: return null
            fun command(response: XmageResponseCommand): GameCommand? {
                val type = response.type ?: return null
                val promptId = response.promptId ?: return null
                val confirmed = response.confirmed ?: response.pay ?: return null
                return UniversalPromptResponseCommandBuilder.command(snapshot.id, snapshot.bridgeRevision, prompt, type, promptId, prompt.playerId,
                    listOf(if (confirmed) "true" else "false"), pay = response.pay ?: confirmed)
            }
            // XMage's left button (yes) mulligans; the right one (no) keeps.
            val mulligan = command(yes) ?: return null
            val keep = command(no) ?: return null
            return OpeningHandChoice(prompt.message.replace(tags, ""), confirmation.noLabel ?: "Keep", confirmation.yesLabel ?: "Mulligan",
                keep, mulligan, prompt.id)
        }
    }
}
